pkg/kubelet/custommetrics/impl.go

```diff
@@ -0,0 +1,149 @@
+// It is parsed from JSON from the value of the annotation with CustomMetricsAnnotationKey.
+// More info: http://releases.k8s.io/HEAD/docs/proposals/custom-metrics.md#api
+type CustomMetricsEndpoint struct {
+	API   string   `json:"api",omitempty`
+	Path  string   `json:"path",omitempty`
+	Port  int      `json:"port"`
+	Names []string `json:"names"`
+}
+
+func AddCustomMetricsLabelIfRequired(pod *api.Pod, labels map[string]string) {
+	if _, ok := pod.Annotations[CustomMetricsAnnotationKey]; ok {
+		labels["io.cadvisor.metric.prometheus"] = CustomMetricsDefinitionContainerPath
+	}
+}
+
+func MakeCustomMetricsMountIfRequired(
+	pod *api.Pod, container *api.Container, podContainerDir string, maxCustomMetricsPerPod int) (*kubecontainer.Mount, error) {
+	if definition, ok := pod.Annotations[CustomMetricsAnnotationKey]; ok {
+		customMetricsMount, err := makeCustomMetricsMount(
+			container, podContainerDir, pod.Status.PodIP, definition, maxCustomMetricsPerPod)
+		if err != nil {
+			return nil, err
+		}
+		return customMetricsMount, nil
+	}
+	return nil, nil
+}
+
+func makeCustomMetricsMount(container *api.Container, podContainerDir, podIP, definition string, maxCustomMetricsPerPod int) (*kubecontainer.Mount, error) {
+	endpoints, err := getCustomMetricsEndpointsForContainer(container, definition, maxCustomMetricsPerPod)
+	if err != nil {
+		return nil, err
+	}
+	if len(endpoints) == 0 {
+		return nil, nil
+	}
+
+	if err := os.MkdirAll(podContainerDir, 0750); err != nil {
+		return nil, fmt.Errorf("error creating pod-container directory: %v", err)
+	}
+	hostsFilePath := path.Join(podContainerDir, "cadvisor-custom-metrics-prometheus.json")
+	if err := ensureCustomMetricsFile(hostsFilePath, podIP, endpoints); err != nil {
+		return nil, err
+	}
+	return &kubecontainer.Mount{
+		Name:          "k8s-custom-metrics",
+		ContainerPath: CustomMetricsDefinitionContainerPath,
+		HostPath:      hostsFilePath,
+		ReadOnly:      false,
+	}, nil
+}
+
+func getCustomMetricsEndpointsForContainer(container *api.Container, definition string, maxCustomMetricsPerPod int) ([]CustomMetricsEndpoint, error) {
+	var endpoints []CustomMetricsEndpoint
+	err := json.Unmarshal([]byte(definition), &endpoints)
+	if err != nil {
+		return nil, fmt.Errorf("error while parsing custom metrics definition '%s': %v", definition, err)
+	}
+
+	numCustomMetrics := 0
+	for _, endpoint := range endpoints {
+		if len(endpoint.Names) == 0 {
+			return nil, fmt.Errorf("custom metrics endpoint definition cannot contain empty names: %s", endpoint)
+		}
+		numCustomMetrics += len(endpoint.Names)
+	}
+	if numCustomMetrics > maxCustomMetricsPerPod {
+		return nil, fmt.Errorf("too many custom metrics per pod, only %d allowed: %v", maxCustomMetricsPerPod, endpoints)
+	}
+
+	var matching []CustomMetricsEndpoint
+	for _, port := range container.Ports {
+		for _, endpoint := range endpoints {
+			if port.ContainerPort == endpoint.Port {
+				matching = append(matching, endpoint)
+			}
```

Comment:
>  Please note that we may have many containers in a pod and only some of them will expose custom metrics and match.

Ah, I see.  I was thinking more at the pod level -- it would be nice to say "you mention custom metrics on a port that you haven't noted on any of this pod's containers -- did you really mean to do that"?